### PR TITLE
Fixes #3192, allows promise callback to be set to null

### LINKED ISSUE #3192: can't cleanup after JsSetPromiseContinuationCallback
I'd like some method to remove my non-null promise continuation callback from the JavaScriptContext.

However, I get a ``JavaScriptErrorCode.NullArgument`` if I call ``JavaScriptContext.SetPromiseContinuationCallback(null, IntPtr.Zero)``

I presume the argument validation is too strict?

### PR DESCRIPTION
**Pull request by jackhorton**

### CODE REVIEW

**Comment by digitalinfinity** [comment]
@dotnet-bot test Tab Check please

File: lib/Jsrt/ChakraCommon.h [hunk 1]
```diff
@@ -2384,7 +2384,7 @@ typedef unsigned short uint16_t;
     /// </returns>
     CHAKRA_API
         JsSetPromiseContinuationCallback(
-            _In_ JsPromiseContinuationCallback promiseContinuationCallback,
+            _In_opt_ JsPromiseContinuationCallback promiseContinuationCallback,
```
Comment (jackhorton) [inline lib/Jsrt/ChakraCommon.h]: @liminzhu This may need a doc update?

File: bin/NativeTests/JsRTApiTest.cpp [hunk 2]
```diff
@@ -1414,9 +1414,10 @@ namespace JsRTApiTest
     static void CALLBACK PromiseContinuationCallback(JsValueRef task, void *callbackState)
     {
         CHECK(callbackState != nullptr);
-        CHECK(*(void **)callbackState == nullptr);
-        // All the task need to finish async, so we need to save the callback.
-        *(void **)callbackState = task;
+        CHECK(*(JsValueRef *) callbackState == JS_INVALID_REFERENCE);
```
Comment (kfarnung) [inline bin/NativeTests/JsRTApiTest.cpp]: I feel like this was more correct before.  callback state is a void pointer so you don't need to cast it to assign the JsValueRef to it.

File: bin/NativeTests/JsRTApiTest.cpp [hunk 3]
```diff
@@ -1428,15 +1429,15 @@ namespace JsRTApiTest
         REQUIRE(JsSetPromiseContinuationCallback(PromiseContinuationCallback, &callback) == JsNoError);
         REQUIRE(JsRunScript(
             _u("new Promise(") \
-            _u("  function(resolve, reject) {") \
-            _u("      resolve('basic:success');") \
-            _u("  }") \
+            _u("    function(resolve, reject) {") \
```
Comment (kfarnung) [inline bin/NativeTests/JsRTApiTest.cpp]: Was this intentional?  It seems like JS tends to use two spaces for indentation anyway, so I'd rather just leave this as it was unless there's a strong opinion to the contrary.

File: lib/Jsrt/Core/JsrtContextCore.h [hunk 4]
```diff
@@ -163,7 +163,6 @@ class ChakraCoreHostScriptContext sealed : public HostScriptContext
 
     HRESULT EnqueuePromiseTask(Js::Var taskVar) override
     {
-        AssertMsg(false, "jsrt should have set the promise callback");
```
Comment (kfarnung) [inline lib/Jsrt/Core/JsrtContextCore.h]: I feel like this should be ok as well, it only hits in debug builds when the embedder has not added a promise callback, but tries to use Promises in their code?

File: bin/NativeTests/JsRTApiTest.cpp (hunk 3 above)
Comment (jackhorton) [inline bin/NativeTests/JsRTApiTest.cpp]: There was one spacing issue with a brace before and that was just how I ended up fixing it. I can revert it back to two spaces.

File: lib/Jsrt/Jsrt.cpp [hunk 5]
```diff
@@ -2932,15 +2932,13 @@ CHAKRA_API JsIdle(_Out_opt_ unsigned int *nextIdleTick)
...
-    return ContextAPINoScriptWrapper_NoRecord([&](Js::ScriptContext * scriptContext) -> JsErrorCode {
-        PARAM_NOT_NULL(promiseContinuationCallback);
-
-        scriptContext->GetLibrary()->SetNativeHostPromiseContinuationFunction((Js::JavascriptLibrary::PromiseContinuationCallback)promiseContinuationCallback, callbackState);
+    return ContextAPINoScriptWrapper_NoRecord([&](Js::ScriptContext *scriptContext) -> JsErrorCode {
+        scriptContext->GetLibrary()->SetNativeHostPromiseContinuationFunction((Js::JavascriptLibrary::PromiseContinuationCallback) promiseContinuationCallback, callbackState);
         return JsNoError;
     },
-    /*allowInObjectBeforeCollectCallback*/true);
+    /*allowInObjectBeforeCollectCallback*/ true);
```
Comment (kfarnung) [inline lib/Jsrt/Jsrt.cpp]: It's not too distracting here, but it's better to minimize the changes you make when completing a bug fix.  These sorts of formatting changes create more opportunities for merge conflicts and can hide the true spirit of the change.  If we want to fix these, I would do a separate PR to take care of them or at least split them out of the bug fix commit and push two separate commits.

File: lib/Jsrt/Core/JsrtContextCore.h (hunk 4 above)
Comment (jackhorton) [inline lib/Jsrt/Core/JsrtContextCore.h]: I took this out after speaking with Jimmy, because we wanted to add a unit test for this and he said testing asserts was problematic.

Comment (MSLaguana) [inline lib/Jsrt/Core/JsrtContextCore.h]: Please correct me if I'm wrong, but I recall that when I made a change to `nativetests` and was trying to test an error case that had a debug assert when the test case hit it, the assert caused the process to exit, and the test wasn't even marked as failed. My perspective here was that if we are explicitly supporting un-setting the promise callback, then having a promise complete with no completion handler is no longer assert-worthy, but instead a documented error case.

Comment (jackhorton) [inline lib/Jsrt/Core/JsrtContextCore.h]: That's exactly what happened, NativeTests.exe crashes in debug mode with that assert in. In the JSRT overview in the wiki, there is mention that you need to use this function in order to use promises. Should something be added to the JsSetPromiseContinuationCallback documentation as well?

File: lib/Jsrt/ChakraCommon.h (hunk 1 above)
Comment (liminzhu) [inline lib/Jsrt/ChakraCommon.h]: Will do after the change merges

File: bin/NativeTests/JsRTApiTest.cpp [hunk 6]
```diff
@@ -1472,6 +1474,48 @@ namespace JsRTApiTest
...
 
+    void UnsetPromiseContinuation(JsRuntimeAttributes attributes, JsRuntimeHandle runtime)
+    {
+        JsValueRef result = JS_INVALID_REFERENCE, callbackState = JS_INVALID_REFERENCE, exception = JS_INVALID_REFERENCE;
+        JsValueType cbStateType = JsUndefined;
+
+        // script with no promise continuation callback should error
+        REQUIRE(JsRunScript(_u("new Promise((res, rej) => res()).then(() => 1)"), JS_SOURCE_CONTEXT_NONE, _u(""), &result) == JsErrorScriptException);
+        CHECK(result == JS_INVALID_REFERENCE);
+        REQUIRE(JsGetAndClearException(&exception) == JsNoError);
```
Comment (MSLaguana) [inline bin/NativeTests/JsRTApiTest.cpp]: Perhaps also verify that the exception that was reported matches the one that we expect?

File: bin/NativeTests/JsRTApiTest.cpp [hunk 7]
```diff
@@ -1478,17 +1478,18 @@ namespace JsRTApiTest
     {
         JsValueRef result = JS_INVALID_REFERENCE, callbackState = JS_INVALID_REFERENCE, exception = JS_INVALID_REFERENCE;
         JsValueType cbStateType = JsUndefined;
+        const wchar_t *script = _u("new Promise((res, rej) => res()).then(() => 1)");
```
Comment (kfarnung) [inline bin/NativeTests/JsRTApiTest.cpp]: I think you want to use WCHAR since wchar_t is not a consistent size across Linux and Windows.

Comment (kfarnung) [inline bin/NativeTests/JsRTApiTest.cpp]: Oh, hmm, it looks like maybe these tests are currently Windows only?  In that case I guess it doesn't matter, but we should probably get them running cross-platform.

**Comment by kfarnung** [comment]
Meta comment, you'll probably want to squash your changes before the final merge and I'd update the commit message title with a description of the bug and just mention "Fixes <foo>" in the body.

File: bin/NativeTests/JsRTApiTest.cpp (hunk 7 above)
Comment (jackhorton) [inline bin/NativeTests/JsRTApiTest.cpp]: if I convert to WCHAR, should I also convert to JsRun? Or, if making these tests xplat is a goal, we should probably just open an issue to do the conversion all at once.

Comment (kfarnung) [inline bin/NativeTests/JsRTApiTest.cpp]: Honestly this method is a little bit of a mess.  It's Windows only and the header uses wchar_t and the implementation uses "WCHAR".  I'd say just leave it for now.

Comment (digitalinfinity) [inline bin/NativeTests/JsRTApiTest.cpp]: I agree- dealing with making NativeTests xplat is a separate issue, and is already tracked with #2590. I'll poke that issue again. For now, you can leave it as wchar_t, we can clean this up later.
